research/conditional/utils/misc_tools.py

```diff
@@ -55,3 +56,37 @@ def stable_softmax_temperature(x, temperature, dim=-1):
 def entropy(x):
     ent = -torch.sum(x * torch.log(x + 1e-8), dim=-1)
     return ent
+
+
+@define
+class TemperatureScheduler:
+    model: torch.nn.Module
+    steps_until_anneal: int
+    total_steps: int
+    previous_multiplier: float = 1.0
+    current_multiplier: float = 1.0
```

Comment:
uwaga główna: temperatura powinna być ustawiona forced na zero przez wiele, wiele kroków treningowych, w stylu ostatnie 1% może 5% treningu. Szczególnie w sytuacji, gdzie temperatura jest tak czy inaczej uczona, i może nam odskakiwać w górę. To powinno wiązać się z zamrożeniem kontrolera (bo gradienty będą zerowe, albo bliskie zeru, ale Adam je będzie starał się wynormalizować....).

Poza tym, można zamrozić temperaturę na czas annealingu.